File: Common/ReferenceEqualityComparer.vb
Namespace Common

    Public Class ReferenceEqualityComparer(Of T As Class)
        Inherits EqualityComparer(Of T)

        Public Overrides Function Equals(x As T, y As T) As Boolean
            Return x Is y
        End Function

        Public Overrides Function GetHashCode(obj As T) As Integer
            Return Runtime.CompilerServices.RuntimeHelpers.GetHashCode(obj)
        End Function

#Region "Instance Property"
        Private Shared ReadOnly _Instance As ReferenceEqualityComparer(Of T) = New ReferenceEqualityComparer(Of T)()

        Public Shared ReadOnly Property Instance As ReferenceEqualityComparer(Of T)
            Get
                Return _Instance
            End Get
        End Property
#End Region

    End Class

End Namespace

File: Common/ReferenceEqualityComparer.cs
using System.Collections.Generic;

namespace Ks.Common
{
    public class ReferenceEqualityComparer<T> : EqualityComparer<T> where T : class
    {
        public override bool Equals(T x, T y)
        {
            return x == y;
        }

        public override int GetHashCode(T obj)
        {
            return System.Runtime.CompilerServices.RuntimeHelpers.GetHashCode(obj);
        }

        public static ReferenceEqualityComparer<T> Instance { get; } = new ReferenceEqualityComparer<T>();
    }
}
